builtin/method_str.py

```diff
@@ -130,3 +136,196 @@ def Call(self, rd):
+        s => replace(string_val, subst_str, count=-1)
+        s => replace(string_val, subst_expr, count=-1)
+        s => replace(eggex_val, subst_str, count=-1)
+        s => replace(eggex_val, subst_expr, count=-1)
+
+        For count in [0, MAX_INT], there will be no more than count
+        replacements. Any negative count should read as unset, and replace will
+        replace all occurances of the pattern.
+        """
+        string = rd.PosStr()
+
+        string_val = None  # type: value.Str
+        eggex_val = None  # type: value.Eggex
+        subst_str = None  # type: value.Str
+        subst_expr = None  # type: value.Expr
+
+        pattern = rd.PosValue()
+        with tagswitch(pattern) as case:
+            if case(value_e.Eggex):
+                # HACK: mycpp will otherwise generate:
+                #  value::Eggex* eggex_val ...
+                eggex_val_ = cast(value.Eggex, pattern)
+                eggex_val = eggex_val_
+
+            elif case(value_e.Str):
+                string_val_ = cast(value.Str, pattern)
+                string_val = string_val_
+
+            else:
+                raise error.TypeErr(pattern,
+                                    'expected pattern to be Eggex or Str',
+                                    rd.LeftParenToken())
+
+        subst = rd.PosValue()
+        with tagswitch(subst) as case:
+            if case(value_e.Str):
+                subst_str_ = cast(value.Str, subst)
+                subst_str = subst_str_
+
+            elif case(value_e.Expr):
+                subst_expr_ = cast(value.Expr, subst)
+                subst_expr = subst_expr_
+
+            else:
+                raise error.TypeErr(subst,
+                                    'expected substitution to be Str or Expr',
+                                    rd.LeftParenToken())
+
+        count = rd.NamedInt("count", -1)
+        rd.Done()
+
+        if count == 0:
+            return value.Str(string)
+
+        if string_val:
+            if subst_str:
+                s = subst_str.s
+            if subst_expr:
+                # Eval with $0 set to string_val (the matched substring)
+                with ctx_EvalReplace(self.mem, string_val.s, None):
+                    s = self.EvalSubstExpr(subst_expr, rd.LeftParenToken())
+            assert s is not None
+
+            result = string.replace(string_val.s, s, count)
+
+            return value.Str(result)
+
+        if eggex_val:
+            ere = regex_translate.AsPosixEre(eggex_val)
+            cflags = regex_translate.LibcFlags(eggex_val.canonical_flags)
+
+            # Walk through the string finding all matches of the compiled ere.
+            # Then, collect unmatched substrings and substitutions into the
+            # `parts` list.
+            pos = 0
+            parts = []  # type: List[str]
+            replace_count = 0
+            while pos < len(string):
+                indices = libc.regex_search(ere, cflags, string, 0, pos)
```

Comment:
I'm not sure eflags (execution flags) should always be 0 ...  In `builtin/method_str.py` there is some logic around `REG_NOTBOL`, where it's only 0 when `pos == 0` when you're searching multiple times

This affects whether `^` can be found in the middle of the string, which would be a bug

It's a bit of a weird API but it kinda makes sense and I think other regex engines have it

Let me see if I can come up with a test case

Reply:
Hm I guess I was wrong about eflags=0 causing this bug?  I saw you added the test, but it already passes without changing the code?

That surprises me but OK, glad it works :)  I can look into it separately ... curious what exactly would fail

Reply:
Hm OK it looks like it is fine and taken care of by `cflags`

`eflags` seems to be only for `leftMatch()` vs. `search()`, I got confused

But at least we have that test case now!

Looks like this is ready to merge then?